cmd/observer/node_builder/observer_builder.go

```diff
@@ -0,0 +1,915 @@
+	"github.com/onflow/flow-go/state/protocol/blocktimer"
+	storage "github.com/onflow/flow-go/storage/badger"
+)
+
+// ObserverBuilder extends cmd.NodeBuilder and declares additional functions needed to bootstrap an Access node
+// These functions are shared by staked and observer builders.
+// The Staked network allows the staked nodes to communicate among themselves, while the public network allows the
+// observers and an Access node to communicate.
+//
+//                                 public network                           staked network
+//  +------------------------+
+//  | observer 1             |<--------------------------|
+//  +------------------------+                           v
+//  +------------------------+                         +----------------------+              +------------------------+
+//  | observer 2             |<----------------------->| Access Node (staked) |<------------>| All other staked Nodes |
+//  +------------------------+                         +----------------------+              +------------------------+
+//  +------------------------+                           ^
+//  | observer 3             |<--------------------------|
+//  +------------------------+
+
+type ObserverBuilder interface {
+	cmd.NodeBuilder
+
+	// IsStaked returns True if this is an Access Node, False otherwise
+	IsStaked() bool
+}
+
+// ObserverServiceConfig defines all the user defined parameters required to bootstrap an access node
+// For a node running as a standalone process, the config fields will be populated from the command line params,
+// while for a node running as a library, the config fields are expected to be initialized by the caller.
+type ObserverServiceConfig struct {
+	staked                       bool // deprecated
+	bootstrapNodeAddresses       []string
+	bootstrapNodePublicKeys      []string
+	observerNetworkingKeyPath    string
+	bootstrapIdentities          flow.IdentityList // the identity list of bootstrap peers the node uses to discover other nodes
+	NetworkKey                   crypto.PrivateKey // the networking key passed in by the caller when being used as a library
+	supportsPublicFollower       bool              // Observers will support streaming to observers later
+	collectionGRPCPort           uint // deprecated
+	executionGRPCPort            uint // deprecated
+	pingEnabled                  bool
+	nodeInfoFile                 string
+	apiRatelimits                map[string]int
+	apiBurstlimits               map[string]int
+	rpcConf                      rpc.Config
+	ExecutionNodeAddress         string                   // deprecated
+	HistoricalAccessRPCs         []access.AccessAPIClient //deprecated
+	logTxTimeToFinalized         bool
+	logTxTimeToExecuted          bool
+	logTxTimeToFinalizedExecuted bool
+	retryEnabled                 bool
+	rpcMetricsEnabled            bool
+	baseOptions                  []cmd.Option
+
+	PublicNetworkConfig PublicNetworkConfig
```

Comment:
> I did some assessment, the reason I kept them:
> supportsPublicFollower Technically observers can have observers attached later for scaling
> collectionGRPCPort, executionGRPCPort Deprecated indeed
> pingEnabled We are still listening to libp2p public network
> nodeInfoFile still serves the private key for the gRPC proxy endpoint
> apiRatelimits, apiBurstlimits, rpcConf we still have gRPC
> I suggest to deprecate some first in this pull request to make sure we do not break consumers triggering a rollback, and remove them later.

1. supportsPublicFollower: I think you're right
2. pingEnabled: Yes, but the ping engine is actually not used for libp2p. It's currently used by our Access Node to monitor the status of partner nodes. Observers don't need this.
3. nodeInfoFile: I don't think this contains any private key, it's just a JSON file containing metadata about partner nodes so that our Access Nodes can output more detailed reporting data for the ping engine above.
4. apiRatelimits, apiBurstlimits, rpcConf maybe we can keep these

Reply:
I fixed these in separate changelists, to easily catch regressions with tests.
nodeInfofile addressed in: https://github.com/GetElastech/flow-go/commit/7a2bda768208cde129aba26e16579f5def811f50
pingEnabled addressed in: https://github.com/GetElastech/flow-go/commit/643ac905dc16de955a56955820ea8a8f6585d36c